ush/ufsda/yamltools.py

```diff
@@ -0,0 +1,241 @@
+import datetime
+import os
+import re
+from solo.yaml_file import YAMLFile
+from solo.template import TemplateConstants, Template
+from ufsda.misc_utils import isTrue
+
+
+def genYAML(input_config_dict, template=None, output=None):
+    """
+    genYAML(input_config_dict, template=None, output=None)
+
+    generate YAML file based on inpput configuration dictionary,
+    environment variables, and optional template.
+
+    input_config_dict - input configuration dictionary
+    template          - path to template YAML file
+    output            - path to output YAML file
+    """
+    # call the parse_config function to create the final dict
+    config_out = parse_config(input_config_dict, template=template)
+    if not output:
+        output = os.path.join(os.getcwd(), 'genYAML_out.yaml')
+    config_out.save(output)
+
+
+def parse_config(input_config_dict, template=None, clean=True):
+    """
+    parse_config(input_config_dict, template=None, clean=True)
+
+    returns a config dict based on input_config_dict, environment vars,
+    and matches an optional template.
+
+    input_config_dict - input configuration dictionary
+    template          - path to template YAML file
+    clean             - if True, and template not None, removes top level dict keys not in template
+    """
+    if template:
+        # open template YAML twice
+        # once for processing, once for cleaning later
+        config_temp = YAMLFile(template)
+        config_out = YAMLFile(template)
+    else:
+        config_out = YAMLFile(data={})
+    # add input_config_dict vars to config for substitutions
+    config_out.update(input_config_dict)
+    # compute common resolution variables
+    if config_out.get('atm', True):
+        config_out = atmanl_case(config_out)
+    config_out.pop('atm', None)  # pop out boolean variable that will cause issues later
+    # do a first round of includes first
+    config_out = include_yaml(config_out)
+    # pull common key values out to top layer
+    config_out = pop_out_common(config_out)
+    # TODO create some vars in config based on other vars if they exist
+    # calculate time/cycle variables
+    config_out = calc_time_vars(config_out)
+    # now recursively update config dict
+    config_out = update_config(config_out)
+    # clean up if specified
+    if clean and template is not None:
+        config_out = clean_yaml(config_out, config_temp)
+
+    return config_out
+
+
+def atmanl_case(config):
```

Comment:
@RussTreadon-NOAA this function and the other one it calls replace the resolution specific YAML files.

This will use `CASE`, `CASE_ENKF`, `DOHYBVAR`, `layout_x` and `layout_y` from either the env or the input YAML to decide what the geometry values will be. Is this what you had in mind to replace redundant YAML files?

Reply:
Thanks, @CoryMartin-NOAA , for adding this flexibility.